Frozen `LJHFile` object

joefowler opened the pull request:
This demonstrates how we might write the `LJHFile` as a frozen dataclass. Use `LJHFile.open(filename)` instead of the previous `LJHFile(filename)` to construct new instances.

joefowler committed:
Don't run ruff checks on marimo examples

joefowler committed:
Fix style problems found by ruff

joefowler committed:
Add explicit test for fractional LJH records

An LJH file that ends in a partial record should be permitted. (It is, but this tests for it.)

joefowler committed:
Add `LJHFile.reopen_binary(pulsesToSkip,...`

(and test it).

joefowler committed:
Check binary pulse data are correct under reopen

joefowler committed:
Remove excessive warning

joefowler committed:
Replace `LJHFile` with a frozen dataclass

joefowler committed:
Correctly use `LJHFile.open(filename)`

joefowler committed:
My marimo automatically changed these indents

ggggggggg commented on moss/channel.py:
```diff
@@ -573,8 +573,8 @@ def from_off(cls, off) -> "Channel":
             df = self.df.select(pl.col("timestamp").cum_max().alias("timestamp")).with_columns(self.df.select(pl.exclude("timestamp")))
-            print("WARNING: in with_experiment_state_df, timestamp is not monotonic, forcing it to be")
-            print("This is likely a BUG in DASTARD.")
+            # print("WARNING: in with_experiment_state_df, timestamp is not monotonic, forcing it to be")
```
I don't like commenting this out, can we just reword it?

ggggggggg reviewed (commented):
Seems pretty reasonable. I had a few comments so I'll hold off on the approve to see your thoughts on them.

joefowler commented on moss/ljhfiles.py:
```diff
@@ -32,100 +67,74 @@ def __read_header(self, filename):
         return self._mmap[i]["data"]
 
-    def to_polars(self, keep_posix_usec=False):
+    def to_polars(self, keep_posix_usec=False) -> Tuple[pl.DataFrame, pl.DataFrame]:
```
Yep. That would work, wouldn't it?

joefowler committed:
Disallow nonzero first pulse in `LJHFile` memmap

But do allow `.to_polars(first_pulse=N)` to skip the first N pulses in the file when generating a Polars dataframe.

joefowler committed:
Complete docstrings and type annotations

joefowler commented on moss/ljhfiles.py:
```diff
@@ -32,100 +67,74 @@ def __read_header(self, filename):
 
-    def __repr__(self):
-        return f"LJHFile {self.filename}"
+    def reopen_binary(self, first_pulse: int = 0, max_pulses: Optional[int] = None) -> Self:
```
Okay, I decided to take your view, that an `LJHFile` represents the entire file, since there's no clear cost to memory-mapping larger areas. On the other hand, by the same argument, why is `max_pulses` allowed?

Would you support removing this parameter, too, and pushing it into `to_polars()`?

ggggggggg replied:
Maybe rename it to `max_pulses_for_testing` In the past I've used a `max_pulses` argument as a simpler way to test extending the underlying file than actually extending the underlying file.

ggggggggg commented on moss/ljhfiles.py:
```diff
@@ -32,100 +67,74 @@ def __read_header(self, filename):
+        if max_pulses is not None:
+            npulses = min(max_pulses, npulses)
+        offset = self.header_size + first_pulse * self.pulse_size_bytes
+        mmap = np.memmap(self.filename, self.dtype, mode="r", offset=offset, shape=(npulses,))
```
But..... maybe it's better not to lie. Lets see if we actually have problems.

joefowler replied:
I don't see why you want to allow mutability. The memmap itself is basically immutable. There's nothing that can be done with it except to wait for the GC to clean it up.

ggggggggg commented:
I think if we use this feature inside marimo we'll have problems since you
would likely have a reference to the old ljhfile, so gc would never happen.
But maybe this never really goes inside marimo, it's for some small
standalone script that just loops this func and applies recipes and writes
the results, its not for interactive use.

On Fri, Aug 1, 2025 at 11:18 AM Joe Fowler ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In moss/ljhfiles.py
> <https://github.com/ggggggggg/moss/pull/15#discussion_r2248484545>:
>
> >
> -    def __repr__(self):
> -        return f"LJHFile {self.filename}"
> +    def reopen_binary(self, first_pulse: int = 0, max_pulses: Optional[int] = None) -> Self:
> +        current_binary_size = os.path.getsize(self.filename) - self.header_size
> +        full_npulses = current_binary_size // self.pulse_size_bytes
> +        npulses = full_npulses - first_pulse
> +        if max_pulses is not None:
> +            npulses = min(max_pulses, npulses)
> +        offset = self.header_size + first_pulse * self.pulse_size_bytes
> +        mmap = np.memmap(self.filename, self.dtype, mode="r", offset=offset, shape=(npulses,))
>
> I don't see why you want to allow mutability. The memmap itself is
> basically immutable. There's nothing that can be done with it except to
> wait for the GC to clean it up.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ggggggggg/moss/pull/15#discussion_r2248484545>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHTXAOHXVDIM5PRSG6F2HT3LOOOHAVCNFSM6AAAAACC25GBG6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAOBQGAYDQMRRHE>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
